File: algorithms/code/leetcode/lc334_increasing_triplet_subsequence/LC334IncreasingTripletSubsequence.java
package leetcode.lc334_increasing_triplet_subsequence;

/**
 * https://leetcode.com/problems/increasing-triplet-subsequence/
 * #medium
 */
public final class LC334IncreasingTripletSubsequence {
    public boolean increasingTriplet(final int[] nums) {
        if (nums == null || nums.length < 3) {
            return false;
        }
        int[] minBefore = new int[nums.length];
        int[] maxAfter = new int[nums.length];
        minBefore[0] = nums[0];
        maxAfter[nums.length - 1] = nums[nums.length - 1];
        for (int i = 1; i < nums.length - 1; i++) {
            minBefore[i] = Math.min(minBefore[i - 1], nums[i - 1]);
            maxAfter[nums.length - 1 - i] = Math.max(maxAfter[nums.length - i], nums[nums.length - i]);
        }
        for (int i = 1; i < nums.length - 1; i++) {
            if (minBefore[i] < nums[i] && nums[i] < maxAfter[i]) {
                return true;
            }
        }
        return false;
    }
}
package leetcode.lc334_increasing_triplet_subsequence;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public final class LC334IncreasingTripletSubsequenceTests {
    @Test
    public void testEmpty() throws Exception {
        assertFalse(new LC334IncreasingTripletSubsequence().increasingTriplet(new int[] {}));
    }

    @Test
    public void testExample1() throws Exception {
        assertTrue(new LC334IncreasingTripletSubsequence().increasingTriplet(new int[] { 1, 2, 3, 4, 5 }));
    }

    @Test
    public void testExample2() throws Exception {
        assertFalse(new LC334IncreasingTripletSubsequence().increasingTriplet(new int[] { 5, 4, 3, 2, 1 }));
    }

    @Test
    public void testOther() throws Exception {
        assertTrue(new LC334IncreasingTripletSubsequence().increasingTriplet(new int[] { 1, 2, 3, 1, 2, 1 }));
    }

    @Test
    public void testNothing() throws Exception {
        assertFalse(new LC334IncreasingTripletSubsequence().increasingTriplet(null));
        assertFalse(new LC334IncreasingTripletSubsequence().increasingTriplet(new int[] { 0, 1 }));
    }

    @Test
    public void test516() throws Exception {
        assertFalse(new LC334IncreasingTripletSubsequence().increasingTriplet(new int[] { 5, 1, 6 }));
    }

    @Test
    public void test24Minus2Minus3() throws Exception {
        assertFalse(new LC334IncreasingTripletSubsequence().increasingTriplet(new int[] { 2, 4, -2, -3 }));
    }
}

File: algorithms/code/leetcode/lc334_increasing_triplet_subsequence/lc334_increasing_triplet_subsequence.rs
// https://leetcode.com/problems/increasing-triplet-subsequence/
// #medium

pub struct Solution;

impl Solution {
    pub fn increasing_triplet(nums: Vec<i32>) -> bool {
        if nums.is_empty() || nums.len() < 3 {
            return false;
        }
        let mut min_before = vec![0; nums.len()];
        let mut max_after = vec![0; nums.len()];
        min_before[0] = nums[0];
        max_after[nums.len() - 1] = nums[nums.len() - 1];
        for i in 1..(nums.len() - 1) {
            min_before[i] = std::cmp::min(min_before[i - 1], nums[i - 1]);
            max_after[nums.len() - 1 - i] =
                std::cmp::max(max_after[nums.len() - i], nums[nums.len() - i]);
        }
        for i in 1..(nums.len() - 1) {
            if min_before[i] < nums[i] && nums[i] < max_after[i] {
                return true;
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty() {
        assert!(!Solution::increasing_triplet(vec![]));
    }

    #[test]
    fn test_example_1() {
        assert!(Solution::increasing_triplet(vec![1, 2, 3, 4, 5]));
    }

    #[test]
    fn test_example_2() {
        assert!(!Solution::increasing_triplet(vec![5, 4, 3, 2, 1]));
    }

    #[test]
    fn test_other() {
        assert!(Solution::increasing_triplet(vec![1, 2, 3, 1, 2, 1]));
    }

    #[test]
    fn test_nothing() {
        assert!(!Solution::increasing_triplet(vec![0, 1]));
    }

    #[test]
    fn test_5_1_6() {
        assert!(!Solution::increasing_triplet(vec![5, 1, 6]));
    }

    #[test]
    fn test_2_4_minus2_minus3() {
        assert!(!Solution::increasing_triplet(vec![2, 4, -2, -3]));
    }
}
